go-controller/pkg/ovn/address_set/address_set.go

```diff
@@ -82,246 +77,177 @@ func NewOvnAddressSetFactory(nbClient libovsdbclient.Client) AddressSetFactory {
 // ovnAddressSetFactory implements the AddressSetFactory interface
 var _ AddressSetFactory = &ovnAddressSetFactory{}
 
-// NewAddressSet returns a new address set object
-func (asf *ovnAddressSetFactory) NewAddressSet(name string, ips []net.IP) (AddressSet, error) {
-	res, err := newOvnAddressSets(asf.nbClient, name, ips)
-	if err != nil {
-		return nil, err
-	}
-	return res, nil
+// NewAddressSet returns a new object that implements AddressSet
+// and contains the given IPs, or an error. Internally it creates
+// an address set for IPv4 and IPv6 each.
+func (asf *ovnAddressSetFactory) NewAddressSet(dbIDs *libovsdbops.DbObjectIDs, ips []net.IP) (AddressSet, error) {
+	return asf.newOvnAddressSets(ips, dbIDs, false)
 }
 
-// ensureAddressSet ensures the address_set with the given name exists and if it does not creates an empty addressSet
-func ensureOvnAddressSet(nbClient libovsdbclient.Client, name string) (*ovnAddressSet, error) {
-	as := &ovnAddressSet{
-		nbClient: nbClient,
-		name:     name,
-		hashName: hashedAddressSet(name),
-	}
-
-	addrSet := nbdb.AddressSet{
-		Name:        as.hashName,
-		ExternalIDs: map[string]string{"name": name},
-	}
-
-	err := libovsdbops.CreateAddressSets(nbClient, &addrSet)
-	// UUID should always be set if no error, check anyway
-	if err != nil || addrSet.UUID == "" {
-		return nil, fmt.Errorf("failed to create address set %+v: %v", addrSet, err)
-	}
-
-	return as, nil
+// EnsureAddressSet makes sure that an address set object exists in ovn
+// with the given dbIDs.
+func (asf *ovnAddressSetFactory) EnsureAddressSet(dbIDs *libovsdbops.DbObjectIDs) (AddressSet, error) {
+	return asf.newOvnAddressSets(nil, dbIDs, true)
 }
 
-// EnsureAddressSet ensures the address_set with the given name exists. If it exists it returns the set
-// and if it does not exist, creates an empty addressSet and returns it.
-func (asf *ovnAddressSetFactory) EnsureAddressSet(name string) (AddressSet, error) {
-	var (
-		v4set, v6set *ovnAddressSet
-		err          error
-	)
-	ip4ASName, ip6ASName := MakeAddressSetName(name)
-	if config.IPv4Mode {
-		v4set, err = ensureOvnAddressSet(asf.nbClient, ip4ASName)
-		if err != nil {
-			return nil, err
-		}
-	}
-	if config.IPv6Mode {
-		v6set, err = ensureOvnAddressSet(asf.nbClient, ip6ASName)
-		if err != nil {
-			return nil, err
-		}
-	}
+func getDbIDsWithIPFamily(dbIDs *libovsdbops.DbObjectIDs, ipFamily string) *libovsdbops.DbObjectIDs {
+	return dbIDs.AddIDs(map[libovsdbops.ExternalIDKey]string{libovsdbops.AddressSetIPFamilyKey: ipFamily})
+}
 
-	return &ovnAddressSets{nbClient: asf.nbClient, name: name, ipv4: v4set, ipv6: v6set}, nil
+func getAddrSetsName(dbIDsWithoutIPFamily *libovsdbops.DbObjectIDs) string {
+	return dbIDsWithoutIPFamily.String()
```

Comment:
this is a little confusing because the variable name is indicating a state of the objectIDs that is not enforced in the code. I don't really understand why this is being done. I guess the name of the address set does not include the ip family so you dont want to get that back in the name here? If that is the case shouldn't we have tags or something for fields that should not be included in primary ID key? Or is this some unique case?

Maybe you should just called the parameter: dbIDs then copy dbIDs and return dbIDs2.RemoveIDs(libovsdbops.AddressSetIPFamilyKey).String() or something

Reply:
yeah that is a "unique" case: this name is only used for logging and by the `FakeAddressSetFactory`. It is used here as a `name`
```
// ovnAddressSets is an abstraction for ipv4 and ipv6 address sets
type ovnAddressSets struct {
	nbClient libovsdbclient.Client
	// name is based on dbIDs without ipFamily
	name string
	ipv4 *ovnAddressSet
	ipv6 *ovnAddressSet
}
```
this struct contains both ipv4 and ipv6 address sets, so its name shouldn't have ipFamily. I've added some comments for this function, since it is quite confusing.

Reply:
Kinda agree with @trozet here.

From the implementation perspective, shouldn't the method itself ensure that the family is not set?

```
func getOvnAddressSetsName(dbIDs *libovsdbops.DbObjectIDs) string {
	if dbIDs.HasID(libovsdbops.AddressSetIPFamilyKey) {
	   dbIDs = dbIDs.RemoveIDs(libovsdbops.AddressSetIPFamilyKey)
        }
        return dbIDs.String()
}
```

Also if this restriction is imposed for the ovnAddressSets type itself, shouldn't that be done through a constructor for the type?

I know that this is private to the package and certain usage correctness can be assumed within if the user is the actual owner of the type, but it is always better to document & ensure correctness through code if possible.